### PR TITLE
Default tolerant zephyr models

### PR DESCRIPTION
**Pull request by k8culver**
### Feature Implemented
Losing a qubit from the Advantage2 typically renders the models (trained on the Advantage2) unusable. The changes in this PR allow for these models to continue to be used even after a qubit or two has been lost.

**Note: The Advantage2 models in this branch are old and have not been updated to the new models. These models were trained prior to the last qubit loss and therefore should be helpful for testing.**

### AI Generation Disclosure
Copilot was given the [fully_yielded_zephyr_subgraph](https://github.com/dwavesystems/dwave-experimental/blob/main/examples/fully_yielded_zephyr_subgraph.py) file and implemented this guided but mostly unaided. I have reviewed it and tested it. Please review carefully.

### CODE REVIEW

**Comment by jackraymond** [comment]
Proof of concept that this works is already finished, IP approved and discussed publicly [2026-06-24-AQC-Jack Raymond.pptx](https://dwave-my.sharepoint.com/:p:/r/personal/jraymond_dwavesys_com/Documents/2026-06-24-AQC-Jack%20Raymond.pptx?d=wbe623b43e7e3480a91a89ec831e9646c&csf=1&web=1&e=zDWGul) . Codebase is already in public pull request https://github.com/dwavesystems/dwave-experimental/pull/54. Just some differences in the training and archiecture conventions to understand (and test).

**Comment by jackraymond** [comment]
Since this pull request was made there have been some method developments relative to [fully_yielded_zephyr_subgraph](https://github.com/dwavesystems/dwave-experimental/blob/main/examples/fully_yielded_zephyr_subgraph.py). That need to be incorporated

File: src/utils/callback_helpers.py [hunk 1]
```diff
@@ -358,13 +358,24 @@ def generate_model_fig(
...
-    subgraph = greedy_get_subgraph(n_nodes=n_latents, random_seed=random_seed, graph=qpu_graph)
-    _, mapping = get_graph_mapping(subgraph)
+    qpu_topology = qpu.properties["topology"]["type"]
 
-    latent_mapping = [mapping[node] for node in subgraph.nodes()]
+    try:
+        subgraph = greedy_get_subgraph(n_nodes=n_latents, random_seed=random_seed, graph=qpu_graph)
+        _, mapping = get_graph_mapping(subgraph)
+        latent_mapping = [mapping[node] for node in subgraph.nodes()]
+    except Exception:
```
Comment (thisac) [inline src/utils/callback_helpers.py:367]: Specify exception caught, or at least print error message? I assume you're trying to catch a specific error here for when `greedy_get_subgraph` fails for find a subgraph?

File: src/utils/callback_helpers.py [hunk 2]
```diff
@@ -358,13 +358,24 @@ def generate_model_fig(
...
+        if qpu_topology != "zephyr":
+            raise
+        fallback = _try_zephyr_sublattice_fallback(
+            n_nodes=n_latents, qpu_graph=qpu_graph, random_seed=random_seed
+        )
+        if fallback is None:
+            raise
```
Comment (thisac) [inline src/utils/callback_helpers.py:374]: Missing exceptions and messages.

File: src/model_wrapper.py [hunk 3]
```diff
@@ -306,14 +307,25 @@ def step(self, batch: tuple[torch.Tensor, torch.Tensor], epoch: int) -> torch.Te
...
+                except BinaryQuadraticModelStructureError:
+                    self._rebuild_sampler()
+                    samples = self._grbm.sample( # type: ignore
+                        sampler=self.sampler,
+                        prefactor=self.PREFACTOR,
+                        linear_range=self.linear_range,
+                        quadratic_range=self.quadratic_range,
+                        device=spins.device,
+                        sample_params=self.sampler_kwargs,
+                    )
```
Comment (thisac) [inline src/model_wrapper.py:328]: For the sake of not copying this twice, you could:

```suggestion
                for _ in range(2):
                    try:
                        samples = self._grbm.sample( # type: ignore
                            sampler=self.sampler,
                            prefactor=self.PREFACTOR,
                            linear_range=self.linear_range,
                            quadratic_range=self.quadratic_range,
                            device=spins.device,
                            sample_params=self.sampler_kwargs,
                        )
                    except BinaryQuadraticModelStructureError:
                        self._rebuild_sampler()
```

Comment (thisac) [reply src/model_wrapper.py:328]: Same below.

File: src/utils/common.py [hunk 4]
```diff
@@ -122,10 +232,23 @@ def get_sampler_and_sampler_kwargs(
...
+        qpu_topology = qpu.properties["topology"]["type"]
+        if qpu_topology != "zephyr":
+            raise
+        fallback = _try_zephyr_sublattice_fallback(
+            n_nodes=n_latents, qpu_graph=qpu_graph, random_seed=random_seed
+        )
+        if fallback is None:
+            raise
+        mapped_graph, chain_embedding = fallback
+        sampler = FixedEmbeddingComposite(qpu, chain_embedding)
```
Comment (thisac) [inline src/utils/common.py:250]: Should this just be put into a helper function and imported into `callback_helpers.py`?

File: src/utils/common.py [hunk 5]
```diff
@@ -14,9 +14,13 @@
 import random
 from typing import Callable, Literal, Optional
 
+import dwave_networkx as dnx
```
Comment (thisac) [inline src/utils/common.py:17]: Better use `dwave-graphs` here instead, since `dnx` is deprecated.

**Review by thisac — COMMENTED** [review]
Apart for some minor comments, LGTM. Would be good to have someone more familiar with the defect zephyr subgraph embedding implementation to review and approve before merging though.

File: src/utils/common.py (hunk 5 above)
Comment (k8culver) [reply src/utils/common.py:17]: Yeah I have this update locally along with the new implementation of Jacks changes I was just waiting for his [PR](https://github.com/dwavesystems/dwave-experimental/pull/54) to be merged so I can test before making a commit

Comment (k8culver) [reply src/utils/common.py:17]: Although maybe it just worth merging these changes and making a followup PR once his PR is merged? Let me know your thoughts

File: src/utils/callback_helpers.py [hunk 6]
```diff
@@ -358,13 +358,24 @@ def generate_model_fig(
...
+    qpu_topology = qpu.properties["topology"]["type"]
 
-    latent_mapping = [mapping[node] for node in subgraph.nodes()]
+    try:
+        subgraph = greedy_get_subgraph(n_nodes=n_latents, random_seed=random_seed, graph=qpu_graph)
+        _, mapping = get_graph_mapping(subgraph)
+        latent_mapping = [mapping[node] for node in subgraph.nodes()]
+    except Exception:
+        if qpu_topology != "zephyr":
+            raise
```
Comment (jackraymond) [inline src/utils/callback_helpers.py:369]: The new code(s) support all processor topologies.

File: src/model_wrapper.py [hunk 7]
```diff
@@ -352,6 +378,34 @@ def step(self, batch: tuple[torch.Tensor, torch.Tensor], epoch: int) -> torch.Te
...
+        qpu_topology = qpu_sampler.properties["topology"]["type"]
+
+        if qpu_topology != "zephyr":
+            raise RuntimeError(
+                f"Sampler rebuild is only supported for Zephyr QPUs (got '{qpu_topology}')."
+            )
+
+        # EmbeddingComposite finds a valid minor embedding of the GRBM's logical
+        # graph on the current working QPU topology without requiring retraining.
+        self.sampler = EmbeddingComposite(qpu_sampler)
```
Comment (jackraymond) [inline src/model_wrapper.py:404]: Does this not mean that minorminer find_embeddings will be called rather than the higher performance options (find_subgraph, find_labeled_subgraph or quotient_embedding_search?).

File: src/utils/common.py [hunk 8]
```diff
@@ -14,9 +14,13 @@
...
 import networkx as nx
 import torch
+from dwave.experimental.embedding_methods import zephyr_quotient_search
 from dwave.system import DWaveSampler, FixedEmbeddingComposite
+from minorminer import find_embedding
+from minorminer.utils.parallel_embeddings import find_sublattice_embeddings
 
 
 def greedy_get_subgraph(
```
Comment (jackraymond) [inline src/utils/common.py:26]: I don't think this is the best way to define a defect tolerant graph.

File: src/utils/common.py [hunk 9]
```diff
@@ -83,6 +87,112 @@ def greedy_get_subgraph(
 
     return subgraph
 
+
+def _try_zephyr_sublattice_fallback(
```
Comment (jackraymond) [inline src/utils/common.py:91]: Seems like quotient search code is only used as a fall back, but isn't greedy_search guaranteed to succeed?

**Review by jackraymond — COMMENTED** [review]
As far as I can tell a processor subgraph is found, in combination with a 1:1 embedding, greedily as a subgraph of the processor to maximize connectivity. If the greedy search fails, a zephyr-quotient subgraph is sought, but I'm not sure under what conditions this fallback would be invoked on brief inspection.

If the graph_id changes then a new embedding is sought using minorminer find_embedding.

I think we could improve this workflow to better align with the concept I showed at AQC 2026 (much of which was developed after this pull request was created), and to incorporate more solid quotient code pending pull requests https://github.com/dwavesystems/minorminer/pull/292 and https://github.com/dwavesystems/dwave-experimental/pull/54.
